pkg/middleware/stored_session.go

```diff
@@ -89,6 +99,25 @@ func (s *storedSessionLoader) loadSession(next http.Handler) http.Handler {
 // getValidatedSession is responsible for loading a session and making sure
 // that is is valid.
 func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *http.Request) (*sessionsapi.SessionState, error) {
+	var session *sessionsapi.SessionState
+	var err error
+	failed := false
+	attempts := 1 + s.retryAttempts
+	for i := 0; i < attempts; i++ {
+		session, err = s.tryToGetValidatedSession(rw, req)
+		if err == nil {
+			if failed {
+				logger.Printf("Previously invalid session is now valid due to retry. (Retries: %d) \n", i)
+			}
+			break
+		}
+		time.Sleep(s.retryDelay)
+		failed = true
+	}
+	return session, err
+}
+
+func (s *storedSessionLoader) tryToGetValidatedSession(rw http.ResponseWriter, req *http.Request) (*sessionsapi.SessionState, error) {
```

Comment:
I don't think this whole function is what you want to retry right (with the potential extract calls to `refreshSessionIfNeeded`? You really only want to retry a `s.store.Load(req)` + checking validity if another parallel request refreshed the token and resulted in a good session?

This current retry organization reads as if you intend to retry the refresh operation? (which is the opposite of what you want that this PR is trying to fix right?)

Reply:
The problem here is that checking if the session is expired or not depends on the Provider. Currently this check is done via `refreshSessionWithProviderIfNeeded`. Apart from that, I can not simply check the session without consulting the Provider.